File: Data Structures & Algorithms/trapping-rain-water/submission-8.java
class Solution {
    public int trap(int[] height) {
        int left = 0;
        int right = height.length - 1;
        int area = 0;
        int leftMax = 0;
        int rightMax = 0;
        while (left < right) {
            leftMax = Math.max(leftMax, height[left]);
            rightMax = Math.max(rightMax, height[right]);
            if (leftMax < rightMax) {
                area += leftMax - height[left]; 
                left++;
            } else {
                area +=  rightMax - height[right];
                right--;
            }
        }
        return area;

    }
}

File: Data Structures & Algorithms/house-robber/submission-8.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        # n = len(nums)
        # f = [0] * (n+2)
        # for i, x in enumerate(nums):
        #     f[i+2] = max(f[i+1], f[i] + x)
        # return f[-1]
        f0 = 0
        f1 = 0
        for i, x in enumerate(nums):
            f0, f1 = f1, max(f1, f0+x)
        return f1        